### DISCUSSION TITLE
viewing reports only displays 30 matches

### DISCUSSION THREAD

**Discussion by SehamAAlharbi** [jplag/JPlag#676] [Q&A] [answered]
Hello, 

I am new to JPlag and was wondering if there is any way to display more than 30 matches when using the API? my data set contains over 80 Java files and for some reason, the report only shows the top 30 matches.

![report](https://user-images.githubusercontent.com/47760671/191334375-51fb6c52-990f-4446-b451-7f45924e4e7b.jpg)

Thanks

**Comment by tsaglam** [comment, marked as answer]
You can define how many matches are going to be displayed with a CLI parameter or programmatically via the options.
Which version of JPlag are you using? v3.0.0 or older? For 3.0.0 and newer, the CLI argument is `-n`, and for legacy and older, it is `-m`.
In the options, it also depends on the version but should be something like `maximumNumberOfComparisons`.

**Comment by FaizAlam** [comment]
Hi @tsaglam, I wanted to ask a clarification regarding the same topic. When i run jplag against my submissions data without mentioning -m option (whose default value is 20), it generates 30 html match files. But when i re-run jplag with -m 40 it generates 40 html match files. 

I do not understand by default how many number of match files should be generated. Going by the default of -m it should be 20 but sometimes 30 or 31 match files are generated. I am using jplag-2.12.1

**Reply by tsaglam** [reply]
Let us say you have 10 submissions. Thus, there are a maximum of $\frac{10*(10-1)}{2} = 45$ of pairs between those submissions.
To limit how large the report gets, there are two parameters to filter these pairs:

- `-m 30` restricts the report to the top 30 matches.
- `-m 30%` restricts the report to all matches with >30% similarity. How many pass that criteria, depends on the dataset itself.

However, this is **legacy syntax** for the **legacy version**, which has a **bug** and thus displays the incorrect default values.

I recommend using an up-to-date version of JPlag, as we are currently at 5.0.0.

**Reply by FaizAlam** [reply]
Hi Timur, thank you for the prompt response. I tried the above things and now I understand the working much better but there are few things which I'm still stuck at.

1) At this point i cannot update the JPlag version because my codebase is using it since 2020 and it would require complete migration to update JPlag.
2) I have kept the number of submission to be 15, then the number of pairs would be 105. Wherein 56 pairs are in similarity bracket of 90-100% and 49 pairs are in 0-10%. When I run the command to generate plag report without `-m` parameter and I get 
```
Matches displayed:	
30 (Treshold: 100.0%) (average similarity)
30 (Treshold: 100.0%) (maximum similarity)
```
Hence i only have 30 match html files in my result directory.

Can you help me understand why there are only 30 matches being displayed(and only 30 matching html files).
I agree that we can use -m to restrict or relax the number of matches (If I run the command with `-m 35`, it generates 35 match files), is there any way to generate matching files for all the matches found, in this case it could be 56 matches.

Regards,
Faiz

**Reply by tsaglam** [reply]
Probably because of a incorrect default limit for not specifying -m [in the codebase](https://github.com/jplag/JPlag/blob/fbded8822d02c852fdc017e225a3d036fe417285/jplag/src/main/java/jplag/options/Options.java#L79) (should be 20, is set to 30 somewhere). Why not just specify `-m 9999` if you want all matches?

**Reply by FaizAlam** [reply]
I will probably be using -m with dynamic parameter, based on number of code files in the folder. By the way, is there any limit to the number of matches being generated?

**Reply by tsaglam** [reply]
No, there is not really a limit. Theoretically, for 65000 submissions you will come close to Integer.MAX_VALUE = ~2 Billion matches, but I would guess this would take quite some time to process anyways.

**Reply by FaizAlam** [reply]
oh alright, i thought 1000 is the limit. I am generating plag report for 312 submission, which would made upto 48516 pairs. and I am using this command `java -jar /home/ubuntu/Jplag/jplag-2.12.1-SNAPSHOT-jar-with-dependencies.jar -m 80% -l c/c++ -r cppRes -s cpp` to generate report of all the pairs having similarity more than 80%. 

This is how my index file looks like:
![image](https://github.com/jplag/JPlag/assets/21370009/2537d48c-4b98-4732-912b-9ed07d611624)
as you can see there are more than 1569+673 submissions pairs which have more than 80% similarity. So the number of files generated should be this much only, but I only get 1000 match files.

**Reply by tsaglam** [reply]
Maybe this is hardcoded somewhere, but I do not know if that is the case. As previously mentioned, the legacy version of JPlag has some bugs and also the report generation is somewhat messy. In the current version you can easily export more than 1000 matches.

**Reply by FaizAlam** [reply]
I found this in legacy [file](https://github.com/jplag/JPlag/blob/fbded8822d02c852fdc017e225a3d036fe417285/jplag/src/main/java/jplag/options/Options.java#L84). Maybe this is why it limit report generation to 1000 files.

**Reply by tsaglam** [reply]
Ah, good catch, this is then probably responsible.
